File: 154-find-minimum-in-rotated-sorted-array-ii/find-minimum-in-rotated-sorted-array-ii.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function (nums) {
    let first = 0, last = nums.length - 1;

    while (first < last) {
        let mid = Math.floor(first + (last - first) / 2);

        if (nums[mid] > nums[last]) {
            first = mid + 1;
        }
        else if (nums[mid] < nums[last]) {
            last = mid;
        }
        else {
            last--;
        }
    }

    return nums[first];
};

File: 154-find-minimum-in-rotated-sorted-array-ii/find-minimum-in-rotated-sorted-array-ii.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        first, last = 0, len(nums) - 1

        while first < last:
            mid = first + (last - first) // 2

            if nums[mid] < nums[last]:
                last = mid
            elif nums[mid] > nums[last]:
                first = mid + 1
            else:
                last -= 1

        return nums[first]
